Add documentation of QualifiedDo to the user's guide (#492)
tweag/ghc: docs/users_guide/exts/qualified_do.rst

  if such a case is needed ::

    M.do { pat <- u; stmts }  =  u M.>>= \case
      { pat -> M.do { stmts }
      ; _ -> M.fail "…"
      }

  If the pattern cannot fail, then we don't need to use ``M.fail``.  ::

    M.do { pat <- u; stmts }  =  u M.>>= \case pat -> M.do { stmts }

*  The desugaring of ``-XApplicativeDo`` uses ``(M.<$>)``, ``(M.<*>)``,
   and ``M.join`` (after the the applicative-do grouping has been performed) ::

    M.do { (x1 <- u1 | … | xn <- un); M.return e }  =
      (\x1 … xn -> e) M.<$> u1 M.<*> … M.<*> un

    M.do { (x1 <- u1 | … | xn <- un); stmts }  =
      M.join ((\x1 … xn -> M.do { stmts }) M.<$> u1 M.<*> … M.<*> un)

  Note that ``M.join`` is only needed if the final expression is
  not identifiably a ``return``.

*  With ``-XRecursiveDo``, ``rec`` blocks use ``M.mfix`` and ``M.return``: ::

     M.do { rec { x1 <- u1; … ; xn <- un }; stmts }  =
       M.do
       { (x1, …, xn) <- M.mfix (\~(x1, …, xn) -> M.do { x1 <- u1; …; xn <- un; M.return (x1, …, xn)})
       ; stmts
       }

If a name ``M.op`` is required by the desugaring process (and only if it's required!) but the name is

> not in scope, an error message is produced:
> 
>   “Desugaring statement ``<stmt>`` requires ``<M.op>`` which is not in scope”

Reviewer: ```suggestion
not in scope, it is reported as an error.
```

The error text does not match closely the implementation, and it is prone to get outdated if we try to copy it.

> The qualified operations are subject to the same type restrictions as their counterparts with ``-XRebindableSyntax``.

Reviewer: The documentation of RebindableSyntax does not highlight what these restrictions are. I thought I had removed this statement from the proposal. I would omit it here as well.

Author: I've removed it!

    Allow the use of qualified ``do`` notation.

``QualifiedDo`` enables qualifying a ``do`` block with a module name, to control which operations to use for
the monadic combinators that the ``do`` notation desugars to.

> With ``-XQualifiedDo`` is enabled, you can *qualify* the ``do`` notation by writing ``modid.do``, where

Reviewer: ```suggestion
When ``-XQualifiedDo`` is enabled, you can *qualify* the ``do`` notation by writing ``modid.do``, where
```
